src/main/java/io/lettuce/core/protocol/CommandHandler.java

```diff
@@ -837,6 +836,15 @@ private static long nanoTime() {
         return System.nanoTime();
     }
 
+    private void discardReadBytesIfNecessary(ByteBuf buffer) {
+        int bufferUsageRatio = clientOptions.getBufferUsageRatio();
+        if (buffer.writerIndex() * (bufferUsageRatio + 1) >= buffer.capacity() * bufferUsageRatio) {
```

Comment:
This calculation is prone for integer overflow.

I wonder whether it would make sense to make discard more customizable. What do you think about providing a `BiConsumer<ByteBuf, DecodeProgress>` that handles discard? The consumer would accept the buffer and an enum value (whether discard happened during decoding a command/after decoding a command/after decoding a batch) and the implementation can then decide over the time and way of release.

The cleaner consumer would be directly returned by `ClientOptions`

Reply:
Yes, current behavior may cause integer overflow. Division operation will be better here. I use this way to calculate by referring to jvm `-XX:GCTimeRatio` usage.

In my opinion, a integer will make this option configure simply. If we want optimize for cpu, we can set a large integer. And a suitable integer for most scene based on CPU and the size of direct memory.  No matter which side we want to optimize, the precondition should always avoid the out of direct memory error. The `-1` can be introduced for only no more free memory or after batch command decoding to discard bytes, if we want keep current behavior which use memory as more as possible when decoding a command.

If use way of `BiConsumer<ByteBuf, DecodeProgress>`, I worry about that make this behavior configure in a complex way. How do you think of this?  Also, if we really need expose to make user do some customized discard.

Reply:
Okay, then let's shoot for the simple approach. Care to fix the overflow so we can prepare this PR for merging?

Reply:
Yes, I updated the PR, Please take a look again.